### PR TITLE
Merge segmented ways to improve labels and shields

### LINKED ISSUE #951: Segmented roads lead to too high or too low density of shields and names
http://www.openstreetmap.org/?mlat=50.0899&mlon=19.9113#map=14/50.0910/19.9144
przechwytywanie
http://www.openstreetmap.org/?mlat=50.0899&mlon=19.9113#map=15/50.0899/19.9113

http://www.openstreetmap.org/?mlat=50.08948&mlon=19.90827#map=18/50.08948/19.90827

przechwytywanie

Similar to #673 (but with reversed problem) and may be also unfixable.

### PR DESCRIPTION
**Pull request by leijurv**
Fixes #951

When a road is split across many ways that all share `name`, `ref`, and highway tags (such as at admin boundaries, when maxspeed changes, etc), each of these individual ways are labeled independently. This results in too many, or too few, labels and shields on what is really one logical road. As described in #951.

Here's a representative test case that I added to my Postgres. The top road is 1 way. The bottom road has 5 nodes connected by 4 ways.

Before this PR, they rendered quite differently:
Screen Shot 2026-05-18 at 9 54 26 PM

After this PR, they render (almost) exactly the same (asterisk see later):
Screen Shot 2026-05-18 at 9 53 02 PM

Changes proposed in this pull request:
- **Join up linestrings that have identical references and highway types**
- Three datasources will now GROUP BY the set of columns that control their rendering: `roads-text-ref`, `roads-text-name`, `roads-text-ref-minor`. In each case, we collect the ways in each group, and call this magical Postgres incantation: `(ST_Dump(ST_LineMerge(ST_Collect(way)))).geom`. First, `ST_LineMerge` looks at the collection of ways, and fuses anything that shares an exact endpoint. Importantly, if a way just crosses another way (while sharing rendering information), that is insufficient - they have to actually touch _at their endpoint_, which practically means they share an OSM node as an endpoint. There is also a `_link` check that is intentionally preserved so that the linestring building is not "distracted" by junctions and offramps. Even though such an offramp might be rendered no different to the highway, we put all offramps in a separate partition. Otherwise the linestring would be interrupted at every offramp because `ST_LineMerge` stops the linestring whenever it reaches an intersection of 3 or more lines. `ST_LineMerge` is run independently in each group/partition, so most of the time it will return a single coherent linestring. But some of the time it will fail to merge everything. Not a problem, maybe there are two roads with the exact same name without actually touching. In that case, `ST_Dump` brings us back down to those individual linestrings again, it's the opposite of `ST_Collect`.
- Directional arrows were wrong in my first version, this is because `ST_LineMerge` completely forgets about the orientation of each line. I replaced `#roads-text-name::directions` with a new path that is not going to do any merging, `#roads-direction-arrows`. In other words, we had to keep arrows rendered exactly as they were, but to do that they had to be separated from `roads-text-name`.
- I did not change road casing/fill. Technically, it changes a few pixels a teensy bit if you apply the line merging there, because of the rounded corners at the end of each way, but this is overkill in my opinion.

[The PostGIS documentation](https://postgis.net/docs/ST_LineMerge.html) for `ST_LineMerge` has these images that make it clear, blue is the "before" and red is the "after":
st_linemerge01
st_linemerge02

This demonstrates how 1. direction is forgotten 2. when three lines come together, `ST_LineMerge` keeps them separate (now you can imagine why offramps were a problem). You may wonder: why not turn on `directed => true` in our call to `ST_LineMerge`? Because the vast majority of roads are not actually one-way roads, and that option would affect **all** roads in the tile, meaning a change in the node order of the way would cause a break in the labeling for all roads.

<details>
<summary>
Asterisk from earlier
</summary>

In order to get the two roads to render **identically**, I had to expand some bboxes, specifically making the labels and shields use `ST_Expand(!bbox!, 2000)`. You can see where I removed that code here https://github.com/leijurv/openstreetmap-carto/commit/aba93ac770c6b8bb4c84a98e0f12abd4976b4a64 So technically that earlier screenshot is not from this exact code with the regular bbox. This is what it looks like for real, as you can see the shields are close but I could not get them to be identical no matter how hard I tried unless I cheated a little bit and selected a larger area due to tile boundary problems. Forgive me, I wanted the impressive screenshot with a truly identical result 🙏 

Screen Shot 2026-05-18 at 9 55 27 PM
</details>

Test rendering. The main change is way more labels are shown. The shields are a bit different but pretty similar.

`#13/40.7408/-73.9584`
nyc png

`#13/52.2321/20.9672`
warsaw png

`#13/37.7610/-122.4440`
sf png

### CODE REVIEW

**Comment by imagico** [comment]
If it only was that simple we would have addressed #951 a long time ago.

Doing `ST_LineMerge()` on a per-tile basis will lead to universal inconsistencies at metatile edges.  This was mentioned in #951.  And it is visible in your samples as well.

For development efficiency it might be worth considering addressing #596 before #951 (which would, however, likely involve quite a bit of Yak shaving). These two things are not only overlapping in the code they need to modify, some potential approaches to addressing the tile edge consistency issues and possible strategies for dealing with the tile expiry issues of #596 also go hand in hand.

**Comment by dch0ph** [comment]
@leijurv I may be wrong, but your test setup may be giving misleading / suboptimal results. The tile servers by default render 8 x 8 metatiles which are split to reduce tile edge effects and improve efficiency. My understanding is that kosmtik does not do this by default and renders 256 x 256 px tiles. This interferes with shield placement since the shield spacing parameters are > 256 px.

My reading is of #951 is that `avoid-edges` may reduce tile edge problems, but AFAIK has never been tested.

I agree that #596 may be the more sustainable long-term fix, but I do think it's worth exploring how far `ST_LineMerge` can take us. I just think your current test set up seems to be aggravating the tile edge effects.

**Comment by leijurv** [comment]
I apologize if this was overeager. I thought you were somewhat suggesting that #5206 and #951 could be my next step, in https://github.com/openstreetmap-carto/openstreetmap-carto/pull/5221#issuecomment-4390824568 (and https://github.com/openstreetmap-carto/openstreetmap-carto/issues/4748#issuecomment-4388117210)

Just to clarify, if there was a regular way that crossed a metatile edge, it would be possibly inconsistent already? Your issue is that when we ST_LineMerge, we cause our ways to be significantly longer than otherwise. Like, this PR here does not **cause** inconsistency at metatile edges for long ways, that already existed (rarely), it's just that this PR **will** cause lots more long ways to exist, which **then** causes more visible metatile inconsistencies?

So in your view, my comparison images above are not an improvement from the left side to the right side. I think I see what you're talking about: that slice at the bottom of manhattan where can see an ugly tile boundary?

I saw in #951 that there was discussion of avoid-edges which [at the time was not available](https://github.com/openstreetmap-carto/openstreetmap-carto/issues/951#issuecomment-56253944) but we can use it now. I have pushed another commit https://github.com/openstreetmap-carto/openstreetmap-carto/pull/5229/changes/1e5e40f3c235b05a90cc1868e4e329abbdcd0edd
 that uses that feature: `text-avoid-edges: true` added to every label symbolizer in `#roads-text-name`/`#roads-text-ref-minor` and `shield-avoid-edges: true` to the shield blocks in `#roads-text-ref`, so that labels and shields near a metatile boundary either fit entirely inside one metatile or are suppressed there and re-placed by the next. I actually had this implemented originally, but I had changed my mind and left it out of this PR - that was a mistake sorry.

Please see the following. You can see the effect of avoid edges: Kent Avenue and Wythe Avenue on the shore of Brooklyn. Previously those labels were chopped up. Now, they are placed cleanly above and below the edge. It is maybe suboptimal that the label is repeated so soon. But that's clearly better than how it was chopped up in my original screenshot.
nyc

**Comment by leijurv** [comment]
@dch0ph Thank you for the suggestion. I have changed `kosmtik serve project.mml --host 0.0.0.0` to `kosmtik serve project.mml --host 0.0.0.0 --metatile 8` to better match osm.org. It did indeed improve the comparison, for example the large "Manhattan Island" label now matches on both sides. However, unluckily, the metatile boundary at the bottom of Manhattan is still present.

<details>
<summary>That NYC comparison on my latest commit <b>with</b> avoid-edges</summary>

https://github.com/openstreetmap-carto/openstreetmap-carto/commit/1e5e40f3c235b05a90cc1868e4e329abbdcd0edd
nyc2
</details>

<details>
<summary>That NYC comparison on the prior commit <b>without</b> avoid-edges</summary>

https://github.com/openstreetmap-carto/openstreetmap-carto/pull/5229/changes/aba93ac770c6b8bb4c84a98e0f12abd4976b4a64
nyc3
</details>

**Comment by imagico** [comment]
I would advise against increasing the metatile size in test rendering.  This would just make the tile edge consistency problems more difficult to spot and analyze.

For better understanding: There are two different types of tile edge inconsistencies:

* inconsistencies resulting from interference cascades between blocking elements (labels and symbols), which - due to different sets of blocking elements in different tiles - are not consistent across tile edges.  These are unavoidable in principle unless you universally avoid rendering of any blocking elements overlapping the tile edges.  The distance across which such cascades can occur is not limited - hence rendering buffers cannot reliably prevent these.  At the same time there are relatively rare to occur practically, especially in cases where blocking elements are not too dense (so the cascades do not get that long).
* inconsistencies resulting from actual differences in the geometries from which blocking symbols and labels are generated in neighboring tiles even if no blocking happens at all.  This is what we have here when you use `ST_LineMerge()` on a per-tile basis.  You can avoid the inconsistencies being visible at the tile edges **on the line merged features themselves** by avoiding labels/symbols crossing the tile edges, but any blocking elements with lower priority than those placed on the line merged geometries can still lead to tile edge inconsistencies - unless you avoid placement on the tile edges for all of those as well.

See also [here](https://imagico.de/blog/en/on-non-locality-in-tiled-rule-based-map-rendering/) for more background info.

Bottom line is: *avoid-edges* is the *give up and use tables* of tile based map rendering :wink:.  I am not saying it can't be considered here, but it needs to be carefully considered in light of what it means practically.

**Comment by leijurv** [comment]
I understand now, thank you for explaining. I was familiar with your first kind of inconsistency, but I failed to realize that second kind of inconsistency.

This PR **would** actually cause a new kind of inconsistency. Previously, if a given way crossed a metatile boundary, both metatiles that have the way would see "the same way". Therefore, both metatiles would, ideally, render a label on that way in the exact same location. When we crop that label down at the metatile boundary, then later attempt to view the label "reassembled" with the metatiles displayed side by side, we would, ideally, see a correctly reassembled label with no seam. This would happen because mapnik would choose to put the label in the exact same place **because** both metatiles "saw" the exact same geometry for that way, so the label decision was, ideally, the exact same. But with this PR, where we assemble ways that are within the current metatile bbox, now the two metatiles have a diverging view of the way. We could imagine a road that has 5 distinct ways, ABCDE, where C crosses the boundary. The left metatile sees a linestring of ABC, the right metatile sees a linestring of CDE, and they'll choose to put a label in probably different places. When we crop to bbox, then reassemble, the label is most likely cut along that edge.

This is a judgement call and I can't really make an argument beyond subjective preference. In my opinion, that latest Manhattan screenshot is improved from the left side to the right side. I do see the downside where those two roads have a label that is duplicated very nearby, on either side of a metatile boundary. And I do see the downside where the metatile boundary also may prevent a label entirely. Finally, I also see the downside that avoiding the edges is an unpleasant solution that is more or less just giving up on the problem of metatile consistency. In my personal opinion, these downsides are worth the upside of drawing significantly more road labels.

My argument is just this: if we assemble ways like this, but we don't place a label across a tile boundary, I would still say the number of "gaps"/"seams" that we won't place a label across is significantly net decreased. I think the "new" labels added are vastly more numerous than the old labels that are now deleted due to spanning two metatiles. My evidence is that in the NYC example (and I can post more if you like), I personally count about 75 additional street labels, at the cost of 1 street label being lost to the tile boundary (Williamsburg Bridge is gone).

**Comment by leijurv** [comment]
I would also say that I think it's a good idea to turn on avoid-edges for labels and shields, in general. Independent of merging segmented ways or not. So if you end up not merging this PR I would separately make another PR to suggest just turning on avoid-edges in a few places.

This is current osm-carto `master` at `#11/34.0993/-118.3289`:
Screenshot 2026-05-21 at 5 33 52 PM

There are more shields now that the refs are hidden, and also local kosmtik has smaller metatiles than osm.org, and LA has a lot of motorways, so this is a worst case scenario for shields getting ripped.

With this change:

```
diff --git a/style/roads.mss b/style/roads.mss
index 3ca7c9c3..19dbb771 100644
--- a/style/roads.mss
+++ b/style/roads.mss
@@ -3713,6 +3713,7 @@ tertiary is rendered from z10 and is not included in osm_planet_roads. */
     shield-margin: @shield-margin;
     shield-face-name: @shield-font;
     shield-clip: @shield-clip;
+    shield-avoid-edges: true;
 
     [highway = 'motorway'] {
       shield-fill: @motorway-shield;
```

It now looks like this:

Screenshot 2026-05-21 at 5 50 24 PM

**Comment by imagico** [comment]
Introducing `avoid-edges` independent of a solution to #951 would be a different matter.  But doing so specifically for the road shields (and possibly road labels) would require a good reason - both in terms of why tile edge inconsistencies are a particular problem with those layers and in terms of the benefits outweighing the disadvantages.

Your example looks odd in the *before* case with nearly all of the tile edge crossing shields being inconsistent across the tile edge.  The most likely explanation is that your test rendering is done without a buffer.  I don't have the impression that on the map on osm.org the shields have a higher occurrence of tile edge inconsistencies than other blocking features.  If that impression is wrong - either generally or at z11 after #5204 - we would need to investigate the causes.  There is no obvious reason why that should be the case.

Please also keep in mind that #951 is not just about increasing the density of shields - it is about shield density being both too high and too low depending on circumstances.  If it was just universally too low we could just adjust the CartoCSS parameters of the shields.  And if you'd introduce `avoid-edges` on the shields that increases the likeliness that you get shields rendered near the tile edge on neighboring tiles on both sides resulting in a weirdly dense shield display for the map user without a reason being visible.  Or in other words: introducing `avoid-edges` on the shields without any other changes would not only not solve #951, it would likely make it worse.

**Comment by dch0ph** [comment]
> Your example looks odd in the _before_ case with nearly all of the tile edge crossing shields being inconsistent across the tile edge. The most likely explanation is that your test rendering is done without a buffer. I don't have the impression that on the map on osm.org the shields have a higher occurrence of tile edge inconsistencies than other blocking features.

Yes, the _before_ case is definitely odd. From digging around, it looks like `mod_tile` sets a `buffer-size` of 128 px if none is set in the XML file. Currently we don't specify a `buffer-size` in the `Map` section of the XML. Arguably we should. 

The `mapnik-reference` [documentation](https://mapnik.org/mapnik-reference/#4.2.0/buffer-size) for `buffer-size` states:
> Extra tolerance around the map (in pixels) used to ensure labels crossing tile boundaries are equally rendered in each tile (e.g. cut in each tile). Not intended to be used in combination with "avoid-edges".

The wording is typically unclear, but this is presumably because `avoid-edges` is doing the same thing, but by shrinking the area in which shields can be placed. It is unclear to me what exactly `avoid-edges` does (is there a fixed margin?).

> I would advise against increasing the metatile size in test rendering. This would just make the tile edge consistency problems more difficult to spot and analyze.

Unfortunately I think we do need to consider / optimise shield placement / density with a specific metatile (and `buffer-size`) in mind.

> @shield-spacing: 760;
> @shield-repeat-distance: 400;
> @shield-margin: 40;

Without metatiles (and no buffer) size [`kosmtik` default], mapnik is rendering a 256 x 256 tile and so the `shield-spacing` parameter surely becomes irrelevant. [The role of the 40 px `shield-margin` is unclear to me.]

Of course, in the current situation, the ways are often fragmented and so the `760 px` has probably little effect. 

The ideal scenario is presumably that shields are placed on a joined up network roughly even spaced by the "shield spacing". 760 px seems rather high to me (we have adjusted the waterway repeat spacing to 500 px). But, as noted, this parameter is currently unlikely to be used much. We currently seem very far from this ideal scenario which would give the desired even shield density.

Note there is an additional parameter:
> shield-label-position-tolerance: 
 Allows the shield to be displaced from its ideal position by a number of pixels (only works with placement:line).

> Default: "shield-spacing/2.0" (If a shield cannot be placed then the renderer will advance by shield-spacing/2.0 to try placement again.)

If this is to be believed (not a given), this means that (currently) a block will result in moving on another 380 px and trying again. This seems a fairly dumb algorithm (compared to searching locally for an unblocked placement). Reducing this parameter might give more sensible results.

**Comment by leijurv** [comment]
I agree that `avoid-edges` is independent of #951, as [above](https://github.com/openstreetmap-carto/openstreetmap-carto/pull/5229#issuecomment-4514002554). I only brought it up to suggest that turning on `avoid-edges` is worth considering in-and-of-itself but I apologize for the red herring, and you certainly know more than me about the details here, I am not familiar with all the differences between kosmtik "as-is" (in its default configuration here) versus osm.org in production. And I am not sure what the best practice is: should I try to locally match osm.org in terms of metatile size and buffer? Or perhaps the best practice is to match osm.org in terms of buffer, but I should keep my metatile small so that tile inconsistencies are most visible?

> And if you'd introduce avoid-edges on the shields that increases the likeliness that you get shields rendered near the tile edge on neighboring tiles on both sides resulting in a weirdly dense shield display for the map user without a reason being visible.

I am not sure if shields are different than labels in some important way here. But let's say for sake of argument that shield placement and label placement are similar problems.

In that case, I would refer to what I said [here](https://github.com/openstreetmap-carto/openstreetmap-carto/pull/5229#issuecomment-4510931293). It is true that label and shield placement would have a new "gap" at each metatile boundary which **IS** a real downside, **but** it comes with the upside of significantly fewer such gaps overall due to linemerging the ways. As we know, OSM editors will split ways very frequently, for many different reasons, like tagging, or relations. In urban areas I've seen roads frequently split into individual ways, once every few blocks, so that each segment of the road can be correctly placed into public transit route relations. My point is that this is vastly more frequent than metatile boundaries.

Here is a table to summarize what I think of the comparison. I think that on both issues, this PR makes an improvement. Please let me know of these four cells which you disagree with?

|Problem|Currently|With this PR|
|---|---|---|
|Problem of labels/shields being ripped at a tile boundary|**Very rare**, due to buffer and large metatiles, but [possible](https://www.openstreetmap.org/#map=13/37.82002/-122.34478)|**Never**, because Mapnik avoids it. (LA example, NYC example)|
|Problem of labels/shields being duplicated (too frequent), or skipped, due to gaps in where they are allowed to be placed|**Common**, due to gaps being created by many separate osm ways, as described in #951 (NYC example: ~75 labels skipped)|**Rare**, because now the only such gaps are at metatile boundaries (NYC example: 2 labels duplicated, 1 label skipped)|

**Comment by imagico** [comment]
I would like to suggest moving most of the matters discussed here out of this PR to suitable issues - if necessary create new ones.  No matter how this PR is going to develop it will not solve many of the matters discussion here touches.  And future work on these things is going to be harder if it needs to find past discussion in a closed PR.  We have the established practice to discuss the problems that needs solving and potential solutions and their pros and cons in issues and only discuss the specifics of a concrete solution developed on the PRs.

I also would like to clarify my remarks on testing practice.  Working on labeling requires both evaluating and adjusting labeling parameters (which is definitely a good idea to do for road labels and shields - this is related to #951, but it would be worth creating a new issue for that) and testing for tile edge discontinuities (which is also required at least in some cases).  Doing both of these things exclusively together is not very efficient though.  Evaluating and adjusting labeling parameters is best done in non-tiled rendering (like using the kosmtik export function or using nik4) while testing tile edge discontinuities is most efficient with tiled rendering using a small metatile size and a tile buffer size matching those common in production use (usually 128px).  The metatile size will influence frequency of tile edge inconsistencies, but that influence is very small compared to that of the buffer size.  Nothing speaks against also testing with metatile sizes and buffer exactly matching production use.  But it is rather difficult to find and study the generally rare tile edge issues in such a setup.

I can't say from the top of my head what methods are available to configure kosmtik to use a certain buffer size in tiled rendering.  Someone should look into this and share their findings - ideally we should document that somewhere in the repository.

> Here is a table to summarize what I think of the comparison. I think that on both issues, this PR makes an improvement. Please let me know of these four cells which you disagree with?

I have not actually tested the PR yet, but i find all of the assessments plausible.  I do, however, consider this binary look at two specific problems and two options to deal with them too narrow.  Even #951 only looks at one dimension - the density of labels and shields.  Non-confusing and natural labeling of roads, however, has a lot of other dimensions.  Arbitrary blocking labels being placed across tile boundaries will inevitably lead to confusing results because map users don't know practically where a metatile boundary is and will wonder *why is there no label here although there should be* in many cases.  And don't forget that road labels do not have the same design for every road class independent of the secondary tags.  See, for example, #2784.

Again: i am not saying we cannot consider introducing this, but doing so by reducing this to a simple binary problem and a single possible solution and based on this (wrong) premise coming to the conclusion that this is the only prudent decision is not the way to go.

I would like to actually test the proposed change before providing a definitive recommendation how to proceed.  But i can give two preliminary recommendations already:

* for solving #951 i would like us to look into approaches that ensure labels/shields are placed in different tiles based on the same geometries.  There are possible approaches to do that both on the fly and through geometry preprocessing during data import.  I understand that this is more fundamental work and developing any such method to a state of being production ready is going to be substantially more time intensive than what is suggested here.  But #951 has been open for more than ten years so i think it is reasonable to consider the different possible approaches that we know of that could be applied to the problem.
* for using `avoid-edges`, independent of if this is part of a solution to #951 or not i would suggest to do meaningful testing with a suitable tile buffer size at different zoom levels and in different geographic settings regarding the practical effects this has.

**Comment by leijurv** [comment]
I see your position much better now. I'm presenting two significant changes (merging linestrings, and avoid-edges), and I'm claiming that even if those changes independently would have unclear cost vs benefit, when you consider them together (like in my table) they appear to be beneficial. But, this is challenging, even if you believe the table, for several reasons:

1. These are two big changes to osm-carto's behavior. Much more testing would be warranted to look for unintended side effects.
2. There could be alternative solutions, with even fewer tradeoffs, that we have not yet considered or thought of. "Yes or no on this PR" is a false dilemma.

Do you think I should make a new issue to discuss those two changes together? Because my position is that now I am not sure that either change on its own is a clear benefit (given what I have now learned about buffer). Or, should I just write about this in #951 because the combination with avoid-edges is already discussed there?

> map users don't know practically where a metatile boundary is and will wonder why is there no label here although there should be in many cases

I'm sorry I know I have said this a few times but do you not agree that users will also wonder just as much "why there is no label although there should be" when a road is split up into ways? And that there are vastly more way gaps then there are metatile boundaries.

For #2784, I would agree that we would want to carefully test to ensure we group linestrings **strictly only if** the ways would be rendered exactly the same, taking note of edge cases such as tunnel tag. Good point.

I actually agree with you that the ideal solution is to `ensure labels/shields are placed in different tiles based on the same geometries`. If neighboring metatiles "see" the exact same geometry for everything to be rendered, boundary artifacts are minimized in the first place. I attempted to implement this the other day with a recursive CTE, see here: https://github.com/leijurv/openstreetmap-carto/compare/d234294a5e7f701428d6eec12ecc8fe7afee3ad4...04924dc5343bcb9ff6dcc1b93c35d44a697593b9 I think this is roughly correct, but it is egregiously slow to run, several additional seconds to make a Los Angeles Z13 tile (I am not sure if that is fixable). It will "follow" any series of ways with matching tags, and it will stop once it either runs out of ways, or the way splits into a junction of 3 ways. The query would give a consistent result, regardless of which metatile I "started in", but I was still seeing some (different) tile boundary artifacts. I could look back into this but I suspect the recursive CTE is simply a no-go from a performance standpoint, given how it makes so many distinct serial queries to planet_osm_line. And for the alternative solution of preprocessing during import, I am really not sure where to begin, or if that is even workable for osm.org because of how they do incremental updates as mappers make edits.

I would be curious to know what you think of this PR as-is, please let me know if/when you get a chance to test it. 👍 

Finally with respect to shields and avoid-edges "on its own", and with respect to what dch0ph said above, and with respect to picking metatile and buffer size. I admit I'm a bit confused on buffer-size and how it interacts with avoid-edges. I would have imagined that any way that crosses a metatile boundary will **always** be seen by both metatiles, since it must be in both bboxes. But I suppose in my Los Angeles screenshot, it's more like the motorways are many individual small ways. Therefore, as we try to put a shield on tiny ways that are near the boundary, the shield itself becomes much larger (at least, wider) than the way, meaning the neighboring tiles make different decisions. Therefore, if we set a butter-size of at least the width of our shields, the shield placement can only indirectly tear across the boundary...? To be honest I'm confused about this: Is it true that if your buffer-size is larger than the largest shield you might draw, then you are **guaranteed** to never tear a shield across a metatile boundary? Or is it just very rare?

Edit: osm.org has a metatile shield tear [here](https://www.openstreetmap.org/#map=13/37.82002/-122.34478) at Z12 and Z13. But it seems this is exceedingly rare.

At the end of the day I feel still pretty unsure about how to do a "fair" comparison. If we are talking about the tradeoff between metatile boundaries vs node gaps like I was in the above table, then it meaningfully matters how big our metatiles are. (smaller metatiles make the comparison worse). And while we could turn on buffer size 128px by default in this repo, that would have been more accurate to osm.org (meaning I wouldn't have been misled by the LA shield tearing), but it also would have masked the effects of geometry merging. I will just use best judgement and make it clear in each comparison if I have changed either from this repo's default, I suppose.

**Comment by imagico** [comment]
> Do you think I should make a new issue to discuss those two changes together? Because my position is that now I am not sure that either change on its own is a clear benefit (given what I have now learned about buffer). Or, should I just write about this in #951 because the combination with avoid-edges is already discussed there?

#951 is for discussing a specific issue and possible approaches to solve it.  If you envision `avoid-edges` as part of a solution to #951 (like in combination with per tile line merging) then that is the place to discuss it.  If you want to discuss use of `avoid-edges` more broadly beyond a specific approach to address #951 opening a new issue for that is advisable.

> > map users don't know practically where a metatile boundary is and will wonder why is there no label here although there should be in many cases
> 
> I'm sorry I know I have said this a few times but do you not agree that users will also wonder just as much "why there is no label although there should be" when a road is split up into ways? And that there are vastly more way gaps then there are metatile boundaries.

Past report from map users have frequently indicated confusion about missing labels - most frequently polygon labels, but also road labels.  Mappers map things with a name and expect these names to be labeled (in analogy to all the other features of the same class where this is evidently the case)  If a name does not turn up despite there obviously being space they wonder why.

> [...] Therefore, if we set a butter-size of at least the width of our shields, the shield placement can only indirectly tear across the boundary...? To be honest I'm confused about this: Is it true that if your buffer-size is larger than the largest shield you might draw, then you are **guaranteed** to never tear a shield across a metatile boundary? Or is it just very rare?

See the two cases discussed abvove in: https://github.com/openstreetmap-carto/openstreetmap-carto/pull/5229#issuecomment-4501915764

In the second case you will usually not have consistency across the tile boundaries.  In the first case inconsistencies will occur across arbitrary distances (so are not ensured to be fully avoided by buffered rendering) due to cascading blocking between elements.  If the buffer size is smaller than half the individual shield/label dimensions (or more generically: smaller than the inherent non-locality of the design elements in the map) there is a third class of tile edge inconsistencies as a result of missing data (unless you diligently use `ST_Expand(!bbox!, ...)` in the layer queries to compensate).

For completeness: There is a fourth class of inconsistencies resulting from inappropriate use of geometry clipping on the CartoCSS level.  But i don't think we have any cases of that left in the style.

> Edit: osm.org has a metatile shield tear [here](https://www.openstreetmap.org/#map=13/37.82002/-122.34478) at Z12 and Z13. But it seems this is exceedingly rare.

Yes.  You can find such when scanning along metatile edges - but they are rare.  At z13 you often have to look across many hundreds of kilometers until you find one.  That does not mean it is not worthwhile to try improving on that.  Incidentally #5227 is already going to do that - less junction labels will mean lower probability of long distance blocking cascades.

**Comment by leijurv** [comment]
I looked more into the [above](https://github.com/openstreetmap-carto/openstreetmap-carto/pull/5229#issuecomment-4514002554) issue, where you [both](https://github.com/openstreetmap-carto/openstreetmap-carto/pull/5229#issuecomment-4520093304) [said](https://github.com/openstreetmap-carto/openstreetmap-carto/pull/5229#issuecomment-4521490492) that my "before" screenshot looked strange. I was using regular OSM-carto `master`. It turns out it is a bug in kosmtik. The !bbox! was being expanded, but the mapnik raster image was not. I have made a PR: https://github.com/kosmtik/kosmtik/pull/364




>>> map users don't know practically where a metatile boundary is and will wonder why is there no label here although there should be in many cases

>> I'm sorry I know I have said this a few times but do you not agree that users will also wonder just as much "why there is no label although there should be" when a road is split up into ways? And that there are vastly more way gaps then there are metatile boundaries.

> Past report from map users have frequently indicated confusion about missing labels - most frequently polygon labels, but also road labels. Mappers map things with a name and expect these names to be labeled (in analogy to all the other features of the same class where this is evidently the case) If a name does not turn up despite there obviously being space they wonder why.

I am not sure if I am miscommunicating, apologies again. I completely understand that missing, or duplicated, labels are a problem. That was the 2nd row of my [table above](https://github.com/openstreetmap-carto/openstreetmap-carto/pull/5229#issuecomment-4521815561), which you found "plausible". My claim is just that the missing labels problem is happening already, today, every time a road way is fragmented because of relations or any other reason. The missing labels problem at metatile boundaries is comparatively many many times less frequent. **If** fixing missing labels on fragmented ways is truly in a direct tradeoff against missing labels at metatile boundaries, then my position is that the former should be considered to outweigh the latter. But perhaps they are not truly in a direct tradeoff, see my last three comments on #951 (https://github.com/openstreetmap-carto/openstreetmap-carto/issues/951#issuecomment-4526397288 and the next two).

**Comment by imagico** [comment]
> 
> [...] My claim is just that the missing labels problem is happening already, today, every time a road way is fragmented because of relations or any other reason. The missing labels problem at metatile boundaries is comparatively many many times less frequent.

Yes, but the splitting of roads in mapping is transparent and intuitive for the mapper.  With some experience they know that if a road is split the labels will not cross the splits.  That this is not good and will often lead to problematic mapping incentives is clear (which is partly why we have #951).  But the difference is that the arbitrary dropping of labels at metatile boundaries is completely non-intuitive for mappers.

>  If fixing missing labels on fragmented ways is truly in a direct tradeoff against missing labels at metatile boundaries, then my position is that the former should be considered to outweigh the latter.

I understand that.  But as i tried to point out here and in #951 - it is not.  There are other possible options to approach #951.  If and when we evaluate those and come to the conclusion that they have unavoidable and unacceptable drawbacks then we can consider this approach.  But i don't want us - for a >10 year old issue - to focus exclusively on an idea that has evident drawbacks and not even seriously look at alternatives.

**Comment by leijurv** [comment]
Okay, understood.

**Comment by leijurv** [comment]
I will close this PR because this approach as-written is very probably not going to be used. Further discussion is in #5252 and #951.

**Comment by reneleonhardt** [comment]
Thank you so much for trying, sad that nothing of your work has been integrated!

**Comment by leijurv** [comment]
@reneleonhardt No it's fine, don't worry, there are still two approaches that are possible! There is aggregating in quadrants, as described in the last few comments of #951, which will be unblocked by #5274, this one is very possible. And there is globally aggregating ahead of time, as described in #5252 (and previously #5244), this one is challenging. You can see what these two approaches look like by expanding the screenshots in [this](https://github.com/openstreetmap-carto/openstreetmap-carto/issues/951#issuecomment-4654637908) comment.
